staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go

```diff
@@ -440,82 +519,97 @@ func (td *typeDiscoverer) discoverStruct(t *types.Type, fldPath *field.Path) ([]
 				for _, tagVal := range tagVals {
 					fakeComments := []string{tagVal}
-					// Extract any embedded list-validation rules.
-					if validations, err := td.validator.ExtractValidations(fmt.Sprintf("%s[vals]", memb.Name), membValType.Elem, fakeComments); err != nil {
-						return nil, fmt.Errorf("%v: %w", fldPath, err)
+					// Extract any embedded key-validation rules.
+					if validations, err := td.validator.ExtractValidations(fmt.Sprintf("%s[keys]", memb.Name), childType.Key, fakeComments); err != nil {
+						return fmt.Errorf("%v: %w", childPath.Key("keys"), err)
 					} else {
 						if len(validations) > 0 {
-							klog.V(5).InfoS("  found list-validations", "n", len(validations))
-							child.eachVal = append(child.eachVal, validations...)
+							klog.V(5).InfoS("  found key-validations", "n", len(validations))
+							child.keyValidations = validations
 						}
 					}
 				}
 			}
-		case types.Slice, types.Array:
 			//TODO: also support +k8s:eachVal
 			if tagVals, found := gengo.ExtractCommentTags("+", memb.CommentLines)[eachValTag]; found {
 				for _, tagVal := range tagVals {
 					fakeComments := []string{tagVal}
 					// Extract any embedded list-validation rules.
-					if validations, err := td.validator.ExtractValidations(fmt.Sprintf("%s[vals]", memb.Name), membValType.Elem, fakeComments); err != nil {
-						return nil, fmt.Errorf("%v: %w", fldPath, err)
+					if validations, err := td.validator.ExtractValidations(fmt.Sprintf("%s[vals]", memb.Name), childType.Elem, fakeComments); err != nil {
+						return fmt.Errorf("%v: %w", childPath.Key("vals"), err)
 					} else {
 						if len(validations) > 0 {
 							klog.V(5).InfoS("  found list-validations", "n", len(validations))
-							child.eachVal = append(child.eachVal, validations...)
+							child.elemValidations = validations
 						}
 					}
 				}
 			}
 		}
 
-		// Extract any field-attached validation rules.
-		if validations, err := td.validator.ExtractValidations(name, membValType, memb.CommentLines); err != nil {
-			return nil, fmt.Errorf("field %s: %w", fldPath.String(), err)
-		} else {
-			if len(validations) > 0 {
-				klog.V(5).InfoS("  found field-attached value-validations", "n", len(validations))
-				child.validations = append(child.validations, validations...)
-			}
-		}
-
-		children = append(children, child)
+		fields = append(fields, child)
 	}
-	return children, nil
+
+	thisNode.fields = fields
+	return nil
 }
 
+// getValidationFunctionName looks up the name of the specified type's
+// validation function.
+//
+// TODO: Currently this is a "blind" call - we hope that the expected function
+// exists, but we don't verify that, and we only emit calls into packages which
+// are being processed by this generator. For cross-package calls we will need
+// to verify the target, either by naming convention + fingerprint or by
+// explicit comment-tags or something.
```

Comment:
Does this mean that for k8s/k8s, the generator will work across all groups since the groups are all included as inputs to a single generator run?

Reply:
yes, but a call to something that is not listed as an input or "peer" (we should rename that - it made sense for conversion, not this) will hit fail the lookup and just say "has no validtion"